### ISSUE TITLE
OpenXml validation works slow starting from version 2.11.3

### ISSUE THREAD

**Issue by rubansergey** [OfficeDev/Open-XML-SDK#918]
- Issue with the OpenXml library

We have a customer's Wordprocessing document **with a lot of tracked changes** that hangs after we update OpenXml.

Investigation shows that:
- It hangs when we call OpenXmlValidator's **Validate** method against MainDocumentPart of this document. 
- It works quick on version 2.11.2
- It was broken on version 2.11.3 I have reviewed [changelog ](https://github.com/OfficeDev/Open-XML-SDK/blob/main/CHANGELOG.md)and there is no information that means quality improvement or making algorithm of validation  more complex. Against, there were some performance improvements like removing recursion. **This points to think that there was some side effect in this build that causes such performance impact**.
- Version 2.12.0 has a CancellationToken support which is great, but is more like a workaround for our situation than a fix, as we still want to validate parts as quick as before.

After this innitial investigation I ran main branch with such simple code added against CPU profiler:

```cs
namespace ConsoleApp1
{
    class Program
    {
        static void Main(string[] args)
        {
            using (var doc = WordprocessingDocument.Open(@"d:\xxxxxx\Workspace\Desk\Sample.docx", false))
            {
                var validator = new OpenXmlValidator();
                validator.Validate(doc.MainDocumentPart);
            }
        }
    }
}
```

Profiler shows that currently the slowest method is DocumentFormat.OpenXml.Validation.ValidationTraverser **ValidatingTraverse** 
![image](https://user-images.githubusercontent.com/52368905/114165458-2ea7d700-9935-11eb-825e-2c7063cc63a3.png)

At this point my investigation stops as I do not have a brunch with build before this method starts works slow and this will be out of scope of my investigation, but I think given information should be enough to at least point to the issue.

Also see attached sample document and VS profiller dump file 
[Sample.docx](https://github.com/OfficeDev/Open-XML-SDK/files/6284918/Sample.docx)
[Report20210409-1040.zip](https://github.com/OfficeDev/Open-XML-SDK/files/6284919/Report20210409-1040.zip)

**Comment by twsouthwick** [comment]
Thanks for the report! I'll take a look.

**Comment by twsouthwick** [comment]
I repro'd this. Here's the benchmark difference:

2.11.2:

```
| Method |    Mean |    Error |   StdDev |
|------- |--------:|---------:|---------:|
|   Test | 1.121 s | 0.0110 s | 0.0102 s |
```

2.13.0-beta0002:

* Too slow to continue *

I'll keep tracking it down. Does look like it is something related to the recursion change.

**Comment by twsouthwick** [comment]
This actually ended up being related to how data is stored during the same validation run. The cached value generated in one of the constraints (which was generated by traversing all the elements) had too narrow of a cache key so it was being evaluated many times in your example document.

This is similar to #763 that was fixed in 2.11.3 due to the same issue.

**Comment by twsouthwick** [comment]
The validation now experiences the following benchmark:

```
| Method |    Mean |    Error |   StdDev |
|------- |--------:|---------:|---------:|
|   Test | 1.377 s | 0.0125 s | 0.0117 s |
```

A little slower, but way better than it was!
